src/libraries/Native/Unix/System.Security.Cryptography.Native.Apple/pal_signverify.c

```diff
@@ -235,59 +92,190 @@ static int32_t ConfigureSignVerifyTransform(SecTransformRef xform,
+            case PAL_SHA384: return kSecKeyAlgorithmRSASignatureDigestPKCS1v15SHA384;
+            case PAL_SHA512: return kSecKeyAlgorithmRSASignatureDigestPKCS1v15SHA512;
         }
-
-        if (!SecTransformSetAttribute(xform, kSecDigestTypeAttribute, cfHashName, pErrorOut))
+    }
+    // Requires macOS 10.13+ or iOS 11+
+    /*if (signatureAlgorithm == PAL_SignatureAlgorithm_RSA_Pss)
+    {
+        switch (hashAlgorithm)
         {
-            return 0;
+            case PAL_SHA1: return kSecKeyAlgorithmRSASignatureDigestPSSSHA1;
+            case PAL_SHA256: return kSecKeyAlgorithmRSASignatureDigestPSSSHA256;
+            case PAL_SHA384: return kSecKeyAlgorithmRSASignatureDigestPSSSHA384;
+            case PAL_SHA512: return kSecKeyAlgorithmRSASignatureDigestPSSSHA512;
         }
+    }*/
+    if (signatureAlgorithm == PAL_SignatureAlgorithm_RSA_Raw)
+    {
+        return kSecKeyAlgorithmRSASignatureRaw;
+    }
 
-        if (hashSize != 0)
-        {
-            CFNumberRef cfHashSize = CFNumberCreate(NULL, kCFNumberIntType, &hashSize);
+    return NULL;
+}
+
+int32_t AppleCryptoNative_SecKeyCreateSignature(SecKeyRef privateKey,
+                                                uint8_t* pbDataHash,
+                                                int32_t cbDataHash,
+                                                PAL_HashAlgorithm hashAlgorithm,
+                                                PAL_SignatureAlgorithm signatureAlgorithm,
+                                                CFDataRef* pSignatureOut,
+                                                CFErrorRef* pErrorOut)
+{
+    if (pErrorOut != NULL)
+        *pErrorOut = NULL;
+    if (pSignatureOut != NULL)
+        *pSignatureOut = NULL;
+
+    if (privateKey == NULL || pbDataHash == NULL || cbDataHash < 0 ||
+        pErrorOut == NULL || pSignatureOut == NULL)
+    {
+        return kErrorBadInput;
+    }
 
-            if (cfHashSize == NULL)
+    int32_t ret = kErrorSeeError;
+    CFDataRef dataHash = CFDataCreateWithBytesNoCopy(NULL, pbDataHash, cbDataHash, kCFAllocatorNull);
+    if (dataHash == NULL)
+        return kErrorUnknownState;
+
+    if (signatureAlgorithm == PAL_SignatureAlgorithm_DSA)
+    {
+#if !defined(TARGET_MACCATALYST) && !defined(TARGET_IOS) && !defined(TARGET_TVOS)
+        SecTransformRef signer = SecSignTransformCreate(privateKey, pErrorOut);
+
+        if (signer != NULL)
+        {
+            if (*pErrorOut == NULL)
             {
-                return 0;
+                if (ConfigureSignVerifyTransform(signer, dataHash, pErrorOut))
+                {
+                    ret = ExecuteSignTransform(signer, pSignatureOut, pErrorOut);
+                }
             }
 
-            if (!SecTransformSetAttribute(xform, kSecDigestLengthAttribute, cfHashSize, pErrorOut))
+            CFRelease(signer);
+        }
+#else
+        return kPlatformNotSupported;
```

Comment:
This is skipping cleanup code.

```suggestion
        ret = kPlatformNotSupported;
```